### ISSUE TITLE
Replicate displays that allow replication of published results

### ISSUE THREAD

**Issue by MichaelCurrie** [openworm/open-worm-analysis-toolbox#91]
(Originally described by Jim [in this comment](https://github.com/JimHokanson/SegwormMatlabClasses/issues/10#issuecomment-58836690))

**Comment by MichaelCurrie** [comment]
Discussing this with Chris Linzy yesterday, it's clear that charts a,b, and c will not be possible to replicate with only the 15-minute videos of adult worms.  Those charts must have been produced with longer videos of growing worms, which we don't currently have, and also which is out of scope of the current project.

The other histograms are possible to be seen with the [generate_stats.py](https://github.com/openworm/movement_validation/blob/master/examples/generate_stats.py) example.

Still to be done: re-do the generate_stats example as an IPython notebook example, just like we're working on for features with [issue 97](https://github.com/openworm/movement_validation/issues/97).

**Comment by MichaelCurrie** [comment]
The heatmap could be created with seaborn, which is a package that sits on top of matplotlib:

http://stanford.edu/~mwaskom/software/seaborn/examples/network_correlations.html

![image](https://cloud.githubusercontent.com/assets/5267078/8367289/ea63fdbc-1b59-11e5-88dd-5aa955f820ad.png)

```Python
import seaborn as sns
import matplotlib.pyplot as plt
sns.set(context="paper", font="monospace")

df = sns.load_dataset("brain_networks", header=[0, 1, 2], index_col=0)
corrmat = df.corr()

f, ax = plt.subplots(figsize=(12, 9))
sns.heatmap(corrmat, vmax=.8, linewidths=0, square=True)

networks = corrmat.columns.get_level_values("network")
for i, network in enumerate(networks):
    if i and network != networks[i - 1]:
        ax.axhline(len(networks) - i, c="w")
        ax.axvline(i, c="w")

f.tight_layout()
```

**Comment by MichaelCurrie** [comment]
Currently we show our plots in a an application window, using matplotlib's standard backend.  For instance:
```
import matplotlib.pyplot as plt

# Simple plot example using matplotlib.
plt.plot([3,1,4,1,5], 'ks-', mec='w', mew=5, ms=20)
# Plot to a window using "vanilla" matplotlib:
plt.show()
```

Having done some research, there are basically three options for showing plots like this in HTML:

**1. matplotlib with an mpld3 "backend":**
```
import matplotlib.pyplot as plt

# Simple plot example using matplotlib.  The plot is not shown yet.
plt.plot([3,1,4,1,5], 'ks-', mec='w', mew=5, ms=20)

import mpld3
mpld3.show()
```
***2. matplotlib with a plotly "backend"***
```
import matplotlib.pyplot as plt

# Simple plot example using matplotlib.  The plot is not shown
# but is represented by mpl_fig_obj
mpl_fig_obj = plt.figure()
plt.plot([3,1,4,1,5], 'ks-', mec='w', mew=5, ms=20)

import plotly.plotly as py
py.sign_in('mcurrie', 'INSERT-API-HERE from https://plot.ly/settings/api')
py.plot_mpl(mpl_fig_obj)
```
**3. Using plotly's own API**
*[I haven't explored precisely how to do this, since plotly is not open-source, and hosts plots on its own servers.*

I've started using the mpld3, starting with e984917deacfe0e1437b0d7bda51e220ea13bf59, which adds this to `plot_eigenworms.py`.

Note, mpld3 does not work for Python 3 under the official release version available through `pip` (see mpld3 issue https://github.com/jakevdp/mpld3/issues/294).

The workaround is to install from source.  Follow the instructions under "Installing from git" in http://mpld3.github.io/install.html#installing-released-versions.

**Comment by MichaelCurrie** [comment]
Possible very bad news: mpld3 doesn't appear to support animation.  Here's a simple example (based on [this gallery example from matplotlib](http://matplotlib.org/1.4.2/examples/animation/basic_example.html)) that just rendered a blank plot window:

```
import numpy as np
import matplotlib.pyplot as plt
import matplotlib.animation as animation
import mpld3

def update_line(num, data, line):
    line.set_data(data[...,:num])
    return line,

fig1 = plt.figure()

data = np.random.rand(2, 25)
l, = plt.plot([], [], 'r-')
plt.xlim(0, 1)
plt.ylim(0, 1)
plt.xlabel('x')
plt.title('test')
line_ani = animation.FuncAnimation(fig1, update_line, 25, fargs=(data, l),
    interval=50, blit=True)

#plt.show()
mpld3.show()
```

**Comment by MichaelCurrie** [comment]
I got the `nbagg` backend (which is derived from the aforementioned `webagg` backend) to show an animation within an IPython notebook, which I'm realizing might be a nice way to show our results anyway.  I've loaded it as [`animation test.ipynb`](https://github.com/openworm/movement_validation/blob/master/examples/animation%20test.ipynb).  You can run it locally on your machine or online via a service like https://jupyter.org.

**Comment by MichaelCurrie** [comment]
Another plot we could show is this lovely one showing path but annotated with omega and piroutte events:

![image](https://cloud.githubusercontent.com/assets/5267078/8483460/05b43a10-20ae-11e5-8555-c36313917cb8.png)

From [Mechanistic analysis of the search behaviour of *Caenorhabditis elegans*] (http://rsif.royalsocietypublishing.org/content/11/92/20131092) by Salvador *et al.*

**Comment by MichaelCurrie** [comment]
I've got the standard NormalizedWormPlottable animation working in an IPython notebook now.  The animation is pretty jerky for me though; it renders only one out of every two to four frames.

![image](https://cloud.githubusercontent.com/assets/5267078/8504620/e3a13466-2191-11e5-876c-dc0d08e6808a.png)

**Comment by MichaelCurrie** [comment]
Another idea is to be able to plot one feature against another in a scatterplot for analysis purposes:

e.g. http://matplotlib.org/examples/axes_grid/scatter_hist.html

![](http://matplotlib.org/_images/scatter_hist.png)

**Comment by MichaelCurrie** [comment]
@clinzy and I are working on this one this week.  `generate_stats_plots.py` is our template for which plots to generate.

**Comment by clinzy** [comment]
The code that was used to generate the original files seems to be scattered here:
https://github.com/openworm/SegWorm/tree/master/Util

**Comment by clinzy** [comment]
Is there a way to separate the features by worm movement state, e.g. Forward Length-Paused Length-Backward-Length?

**Comment by MichaelCurrie** [comment]
@clinzy [HistogramManager](https://github.com/openworm/movement_validation/blob/master/movement_validation/statistics/histogram_manager.py) does this.  For any given feature it calculates histograms sampled when the worm is moving forward, backward, and paused.

**Comment by clinzy** [comment]
Currently there is code to to generate a PDF similar to those generated by Shaffer Labs for histograms of features such as length, speed, etc. There are additional plots that show the worm paths annotated with turns and color coded for values of various features. This latter part I am currently looking at in terms of interface/design patterns to allow the plots to be generated and referenced similar to the histograms.

**Comment by clinzy** [comment]
I have created classes that plot the path of the head, mid, and tail portions as well as the overall worm. I have also created classes to provide point annotations for the turns and start/end positions. There is also a class that colors the worm path based on the velocity at each frame but I am unable to find a set of velocities in the WromFeatures object resulting from the from_disk method that contain valid velocity data. A sample of the current capabilities can be seen in the attached image. 
![test_path_plots](https://cloud.githubusercontent.com/assets/3114194/10547437/2bbb6b9e-73fa-11e5-9c28-ba5e9a082f7b.png)
